### PR TITLE
Addition of interactions to actual visualisations

### PR DESCRIPTION
**Pull request by jujusb**

### CODE REVIEW

File: vis/xAPISG-CompletablesTimes.ipynb [hunk 1]
```diff
@@ -59,33 +68,72 @@
...
     "                    if total_time < minimum:\n",
     "                        minimum = total_time\n",
+    "        if minimum==float('Inf'):\n",
+    "            minimum=maximum\n",
     "        max_times.append(maximum)\n",
     "        min_times.append(minimum)\n",
-    "        \n",
-    "        \n",
+    "    N=len(completablesSelectedCompletableTime)\n",
+    "    n_groups=N\n",
```
Comment (crisal24) [inline vis/xAPISG-CompletablesTimes.ipynb:76]: This assignment is later overwritten (l 105)

File: vis/xAPISG-CompletablesTimes.ipynb [hunk 2]
```diff
@@ -59,33 +68,72 @@
...
+    "    if eyex==0:  #getAllDataInAListOfList\n",
+    "        completablesSelectedCompletableTime=[completablesSelectedCompletableTime]\n",
+    "        min_times=[min_times]\n",
+    "        max_times=[max_times]\n",
+    "        nbtimes=1\n",
+    "    elif eyex==-1:  #getnbFirstData\n",
+    "        nb=10\n",
+    "        nbtimes=1\n",
```
Comment (crisal24) [inline vis/xAPISG-CompletablesTimes.ipynb]: Clarify meaning of variables e.g. changing variable names

File: vis/xAPISG-CompletablesTimes.ipynb [hunk 3]
```diff
@@ -59,33 +68,72 @@
...
+    "        completablesSelectedCompletableTime=[completablesSelectedCompletableTime[:nb]]\n",
+    "        max_times=[max_times[:nb]]\n",
+    "        min_times=[min_times[:nb]]\n",
+    "        if N>nb:\n",
+    "            N=nb\n",
+    "    elif eyex==-2:  #getnbLastData\n",
+    "        nb=10\n",
+    "        nbtimes=1\n",
```
Comment (crisal24) [inline vis/xAPISG-CompletablesTimes.ipynb]: Some assignments could be performed outside if/elif/else for instance this nbtimes=1, which only changes in the last else

File: vis/xAPISG-CorrectIncorrectPerPlayer.ipynb [hunk 4]
```diff
@@ -45,90 +54,54 @@
     "            total_false += results.count(\"False\") +results.count(False)\n",
     "            if total_true+total_false>total_max:\n",
     "                total_max=total_true+total_false\n",
-    "        true_per_player.append(total_true)\n",
-    "        false_per_player.append(total_false)\n",
+    "        true_false_per_player[player][\"true\"]=total_true\n",
+    "        true_false_per_player[player][\"false\"]=total_false\n",
     "    total_max+=0.5\n",
```
Comment (crisal24) [inline vis/xAPISG-CorrectIncorrectPerPlayer.ipynb:67]: Why this?

File: vis/xAPISG-CorrectIncorrectPerQuestion.ipynb [hunk 5]
```diff
@@ -27,81 +29,80 @@
...
-    "                alt_true += results.count(\"True\") +results.count(True) \n",
+    "                alt_true += results.count(\"True\") +results.count(True)\n",
     "                alt_false += results.count(\"False\") +results.count(False)\n",
     "                if alt_true+alt_false>alt_max:\n",
     "                    alt_max=alt_true+alt_false\n",
-    "        true_per_alternative.append(alt_true)\n",
-    "        false_per_alternative.append(alt_false)\n",
+    "        true_false_per_alternative[alt][\"true\"]=alt_true\n",
+    "        true_false_per_alternative[alt][\"false\"]=alt_false\n",
     "    alt_max+=0.5\n",
```
Comment (crisal24) [inline vis/xAPISG-CorrectIncorrectPerQuestion.ipynb]: same as previous comment

File: vis/xAPISG-CorrectIncorrectPerQuestion.ipynb [hunk 6]
```diff
@@ -27,81 +29,80 @@
...
-    "            ylabel=\"percentage of good response\"\n",
+    "            if N>0:\n",
+    "                # Bar chart showing the total number of correct and incorrect alternatives\n",
+    "                # selected by players in each multiple-choice question\n",
+    "                get_two_superposed_values_bars(ind,list_alt,\n",
+    "                               true_per_alternative, \"Correct\",'g',\n",
+    "                               false_per_alternative, \"Incorrect\",'r')\n",
+    "            else:\n",
+    "                noDataToFillVis(10)\n",
+    "            ylabel=\"Number of respose\"\n",
```
Comment (crisal24) [inline vis/xAPISG-CorrectIncorrectPerQuestion.ipynb]: response

File: vis/xAPISG-AlternativesSelectedQuestion.ipynb [hunk 7]
```diff
@@ -5,7 +5,9 @@
    "execution_count": null,
    "outputs": [],
    "source": [
-    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill"
+    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill\n",
+    "%run vis/helpersFunctions/barCharts.ipynb\n",
+    "%run vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb"
```
Comment (crisal24) [inline vis/xAPISG-AlternativesSelectedQuestion.ipynb]: Remove vis/ from path

File: vis/xAPISG-CompletablesProgress.ipynb [hunk 8]
```diff
@@ -5,7 +5,8 @@
    "execution_count": null,
    "outputs": [],
    "source": [
-    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill"
+    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill\n",
+    "%run vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb"
```
Comment (crisal24) [inline vis/xAPISG-CompletablesProgress.ipynb]: Remove vis/ from path

File: vis/xAPISG-CompletablesScores.ipynb [hunk 9]
```diff
@@ -5,7 +5,9 @@
    "execution_count": null,
    "outputs": [],
    "source": [
-    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill"
+    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill\n",
+    "%run vis/helpersFunctions/barCharts.ipynb\n",
+    "%run vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb"
```
Comment (crisal24) [inline vis/xAPISG-CompletablesScores.ipynb]: Remove vis/ from path

File: vis/xAPISG-CompletablesTimes.ipynb [hunk 10]
```diff
@@ -5,7 +5,9 @@
    "execution_count": null,
    "outputs": [],
    "source": [
-    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill"
+    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill\n",
+    "%run vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb\n",
+    "%run vis/helpersFunctions/barCharts.ipynb"
```
Comment (crisal24) [inline vis/xAPISG-CompletablesTimes.ipynb]: Remove vis/ from path

File: vis/xAPISG-CorrectIncorrectPerPlayer.ipynb [hunk 11]
```diff
@@ -8,7 +8,9 @@
    },
    "outputs": [],
    "source": [
-    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill"
+    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill\n",
+    "%run vis/helpersFunctions/barCharts.ipynb\n",
+    "%run vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb"
```
Comment (crisal24) [inline vis/xAPISG-CorrectIncorrectPerPlayer.ipynb]: Remove vis/ from path

File: vis/xAPISG-CorrectIncorrectPerQuestion.ipynb [hunk 12]
```diff
@@ -8,7 +8,9 @@
    },
    "outputs": [],
    "source": [
-    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill"
+    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill\n",
+    "%run vis/helpersFunctions/barCharts.ipynb\n",
+    "%run vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb"
```
Comment (crisal24) [inline vis/xAPISG-CorrectIncorrectPerQuestion.ipynb]: Remove vis/ from path

File: vis/xAPISG-GamesStartedCompleted.ipynb [hunk 13]
```diff
@@ -2,8 +2,16 @@
...
+   "execution_count": 1,
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "ERROR:root:File `'vis/xAPISG-noDataToFillVisualization.ipynb.py'` not found.\n"
+     ]
+    }
+   ],
```
Comment (crisal24) [inline vis/xAPISG-GamesStartedCompleted.ipynb]: Update file not to show error and remove vis/ from path

File: vis/xAPISG-VideosSeenSkipped.ipynb [hunk 14]
```diff
@@ -5,12 +5,15 @@
    "execution_count": null,
    "outputs": [],
    "source": [
-    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill"
+    "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill\n",
+    "%run vis/helpersFunctions/barCharts.ipynb\n",
+    "%run vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb"
```
Comment (crisal24) [inline vis/xAPISG-VideosSeenSkipped.ipynb]: Remove vis/ from path

File: vis/xAPISG-VideosSeenSkipped.ipynb [hunk 15]
```diff
@@ -27,59 +30,88 @@
...
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 2,
    "outputs": [],
    "source": [
-    "def vis_videos_seen_skipped (players_info):\n",
+    "def vis_videos_seen_skipped():\n",
+    "    videos_list=sorted(get_full_list_of_objects(\"videos_seen\",action2=\"videos_seen\", dict=False))\n",
+    "    gamestartcomplete=interactive(construct_graph_video,\n",
```
Comment (crisal24) [inline vis/xAPISG-VideosSeenSkipped.ipynb]: change name 'gamestartcomplete' to current visualizacion (videos seen skipped)

File: vis/helpersFunctions/barCharts.ipynb [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+    "\n",
+    "Input :\n",
+    "* ind : the indicator to know where display the superposed bars\n",
+    "* list_xticks_labels : the list of x_ticks_labels to display\n",
+    "* listUp : the list of values to display up\n",
+    "* labelUp : the labels corresponding of the values to the listUp\n",
+    "* colorUp : the color to the up bars\n",
+    "* listUp : the list of values to display down\n",
+    "* labelUp : the labels corresponding of the values to the listDown\n",
+    "* colorUp : the color to the down bars"
```
Comment (crisal24) [inline vis/helpersFunctions/barCharts.ipynb]: Change name of parameters to listDown, labelDown and colorDown

File: vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb [hunk 17]
```diff
@@ -0,0 +1,341 @@
+{
```
Comment (crisal24) [inline vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb]: Change file name to more clear / descriptive name

File: vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb [hunk 18]
```diff
@@ -0,0 +1,341 @@
...
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "***organizeDictPerEyeXSelected* function** to separate dict to have a certain data to see for the dict of the position of the bars\n",
+    "\n",
+    "Imput :\n",
+    "* listOfKeysDict : the list of the keys of the dictionary\n",
+    "* eyex : 0 if all data, -1 (-2) if 10 first (10 last), or 5, 25, 50 per visualisation for x axis\n",
```
Comment (crisal24) [inline vis/helpersFunctions/OrderAlphaOrValueAndOrganizeDataWithAxe.ipynb]: change name to x axis?

File: widgets/multipleSelectorCheckboxesWidget.ipynb [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "import requests\n",
+    "import random\n",
+    "import ipywidgets as widgets\n",
```
Comment (crisal24) [inline widgets/multipleSelectorCheckboxesWidget.ipynb]: Already imported in xAPISGProcessor

File: widgets/multipleSelectorCheckboxesWidget.ipynb [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+    "        options_widget.children =new_options\n",
+    "    select_any_button.on_click(on_select_any_button)\n",
+    "\n",
+    "\n",
+    "    @output_widget.capture()\n",
+    "    def on_checkbox_change(change):\n",
+    "        selected_recipe = change[\"owner\"].description\n",
+    "        #print(options_widget.children)\n",
+    "        #selected_item = widgets.Button(description = change[\"new\"])\n",
+    "        #selected_widget.children = [] #selected_widget.children + [selected_item]\n",
```
Comment (crisal24) [inline widgets/multipleSelectorCheckboxesWidget.ipynb]: Remove commented code

File: widgets/multipleSelectorCheckboxesWidget.ipynb [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+    "        width='300px',\n",
+    "        height='300px',\n",
+    "        flex_flow='column',\n",
+    "        display='flex'\n",
+    "    )\n",
+    "\n",
+    "    select_all_button = widgets.Button(description='Select All')\n",
+    "    select_any_button = widgets.Button(description='Unselect All')\n",
+    "    select_widget=widgets.HBox([select_any_button, select_all_button])\n",
+    "    #selected_widget = widgets.Box(children=[options[0]])\n",
```
Comment (crisal24) [inline widgets/multipleSelectorCheckboxesWidget.ipynb]: Remove commented code. Also below.

File: widgets/multipleSelectorCheckboxesWidget.ipynb [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+    "                key = lambda x: x.value, reverse = True\n",
+    "            ) #[options_dict[x] for x in close_matches]\n",
+    "        options_widget.children = new_options\n",
+    "\n",
+    "    search_widget.observe(on_text_change, names='value')\n",
+    "    display(output_widget)\n",
+    "    return multi_select\n",
+    "\n",
+    "\n",
+    "#options_dict = {\n",
```
Comment (crisal24) [inline widgets/multipleSelectorCheckboxesWidget.ipynb]: Remove all commented code below.

File: widgets/selectFileWidget.ipynb [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+    }
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "import ipywidgets as widgets\n",
```
Comment (crisal24) [inline widgets/selectFileWidget.ipynb]: Already imported in xAPISGProcessor

File: widgets/selectFileWidget.ipynb [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+    "                else:\n",
+    "                    self.files.append(f)\n",
+    "        \n",
+    "    def widget(self):\n",
+    "        box = widgets.VBox()\n",
+    "        self._update(box)\n",
+    "        return box\n",
+    "    \n",
+    "    def _update(self, box):\n",
+    "          #try:\n",
```
Comment (crisal24) [inline widgets/selectFileWidget.ipynb]: Remove commented code

File: widgets/selectFileWidget.ipynb [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+    "            buttons.append(button)\n",
+    "            self.buttonRun.visible=True\n",
+    "            buttons.append(self.buttonRun)\n",
+    "            self.output.clear_output(wait=False)\n",
+    "            with self.output:\n",
+    "                print(\"Click to run analyse to the file\")\n",
+    "            buttons.append(self.output)\n",
+    "        box.children = tuple([widgets.HTML(\"<h2>%s</h2>\" % (self.path,))] + buttons)\n",
+    "\n",
+    "# example usage:\n",
```
Comment (crisal24) [inline widgets/selectFileWidget.ipynb]: Remove commented code

File: xAPISGProcessor.ipynb [hunk 26]
```diff
@@ -39,321 +39,408 @@
...
+    "**File explorer** to find the file you want to analyse with **Statement Processor** to generate all data and visualisations"
+   ],
+   "metadata": {
+    "collapsed": false,
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   }
   },
   {
```
Comment (crisal24) [inline xAPISGProcessor.ipynb]: From here below, the first three code cells appear again. Please remove duplicates.

File: xAPISGProcessor.ipynb [hunk 27]
```diff
@@ -39,321 +39,408 @@
...
+    "collapsed": false
+   }
   },
   {
    "cell_type": "code",
-   "execution_count": 12,
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "%run ProcessxAPISGStatement.ipynb # notebook to process an xAPI-SG statement\n",
```
Comment (crisal24) [inline xAPISGProcessor.ipynb]: Revise if some parts of this code (and maybe also from previous cells) could be hidden in other notebooks, as we want to keep the main notebook as some higher-level user-friendly dashboard.

**Review by crisal24 — CHANGES_REQUESTED** [review]
Great work! See the comments for some minor issues. 

Some other minor issues with respect to displaying will be commented in private. Also, consider how this PR is to be integrated with changes in other PRs, including the new visualizations.

File: vis/xAPISG-CorrectIncorrectPerPlayer.ipynb (hunk 4 above)
Comment (jujusb) [reply vis/xAPISG-CorrectIncorrectPerPlayer.ipynb:67]: I put all in a dictionary to get in memory the association player, (true, false) response and the value number to order all after with function "orderDict"

File: vis/xAPISG-CorrectIncorrectPerQuestion.ipynb (hunk 5 above)
Comment (jujusb) [reply vis/xAPISG-CorrectIncorrectPerQuestion.ipynb]: I put all in a dictionary to get in memory the association alternative, (true, false) response and the value number to order all after with function "orderDict"

File: globalsSelectors.ipynb [hunk 28]
```diff
@@ -0,0 +1,114 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "source": [
+    "***order* widget* is constructing a dropdown to order for alphabetic keys or for values"
```
Comment (crisal24) [inline globalsSelectors.ipynb]: Revise markdown cells to display correctly

File: globalsSelectors.ipynb [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+    "        x: widgets.Checkbox(\n",
+    "            description=x,\n",
+    "            value=True,\n",
+    "            style={\"description_width\":\"0px\"}\n",
+    "        ) for x in sorted(list(players_info.keys()))\n",
+    "    }\n",
+    "    checkboxesPlayersSelected= multi_checkbox_widget(optionsPlayer)\n",
+    "    buttonRun= widgets.Button(description='Run')\n",
+    "    buttonRun.on_click(displayvis)\n",
+    "    display(VBox([checkboxesPlayersSelected, buttonRun]))\n",
```
Comment (crisal24) [inline globalsSelectors.ipynb]: It is now a little bit less intuitive that we have to click again "Run" button after changing the selected players - could be more user-friendly to direcly see the changes in the dashboard as you select / deselect players, as in the previous version.

File: vis/xAPISG-GamesStartedCompleted.ipynb [hunk 30]
```diff
@@ -16,28 +16,38 @@
...
   },
   {
    "cell_type": "code",
    "execution_count": null,
    "outputs": [],
    "source": [
-    "def vis_games_started_completed(players_info):\n",
+    "def vis_games_started_completed():\n",
+    "    out=interactive(construct_vis_games_started_completed)\n",
+    "    md_text = widgets.HTML(\"#Pie chart with number of games started and completed\\n\"\n",
```
Comment (crisal24) [inline vis/xAPISG-GamesStartedCompleted.ipynb]: Remove #

File: vis/xAPISG-VideosSeenSkipped.ipynb [hunk 31]
```diff
@@ -27,59 +30,90 @@
...
+    "                       order=order,\n",
+    "                       videosSelected=widgets.SelectMultiple(options=videos_list,\n",
+    "                                                             value=videos_list,\n",
+    "                                                             description='Videos',\n",
+    "                                                             disabled=False),\n",
+    "                       percentageVideoSkipped=False,\n",
+    "                       orderLabelValues=[(\"Order by alls seen and skipped videos values\",0),\n",
+    "                                         (\"Order by seens videos values\",\"seen\"),\n",
+    "                                         (\"Order by skipped videos values\",\"skipped\")])\n",
+    "    md_text = widgets.HTML(\"#Pie chart with number of games started and completed\")\n",
```
Comment (crisal24) [inline vis/xAPISG-VideosSeenSkipped.ipynb]: Remove # and change text to actual visualization

File: vis/xAPISG-GamesStartedCompleted.ipynb [hunk 32]
```diff
@@ -16,28 +16,38 @@
...
+    "                       \"Use the playerSelected widget to select the players you want to analyse\")\n",
```
Comment (crisal24) [inline vis/xAPISG-GamesStartedCompleted.ipynb]: This info should appear in all tabs or even above the tabs

**Review by crisal24 — CHANGES_REQUESTED** [review]
Great work! The visualizations and interface have clearly improved a lot!

Some minor issue (besides the comments): when data is not suitable to fill a visualization, sometimes the visualization appears with the "no data to fill" message, while other times, the visualizations does not appear at all, but the widgets to select and configure it (e.g. title, x axis, selection of completable) do appear, which is a bit confusing.

**Comment by jujusb** [comment]
> Great work! The visualizations and interface have clearly improved a lot!
> 
> Some minor issue (besides the comments): when data is not suitable to fill a visualization, sometimes the visualization appears with the "no data to fill" message, while other times, the visualizations does not appear at all, but the widgets to select and configure it (e.g. title, x axis, selection of completable) do appear, which is a bit confusing.

For the widgets, when the visualization appears with the "no data to fill" message, I can't deappear the widgets to select and configure it. For the visualizations that not appear at all,  they normally appear now all the time.

**Review by gorco — APPROVED** [review]

**Pull request merged**